### PR TITLE
Tutorial for mock driver inside python

### PR DESCRIPTION
**Pull request by Anthony25**
Related to https://github.com/napalm-automation/napalm/pull/715

### CODE REVIEW

File: docs/tutorials/mock_driver.rst [hunk 1]
```diff
@@ -1,30 +1,155 @@
-Napalm mock driver
-==================
+Unit tests: Mock driver
+=======================
+
+A mock driver is a software that imitates the response pattern of another
+system. It is meant to do nothing but returns the same predictible result,
+most of the cases in a testing environment.
```
Comment (sieben) [inline docs/tutorials/mock_driver.rst]: usually in a testing environment.

File: docs/tutorials/mock_driver.rst [hunk 2]
```diff
@@ -1,30 +1,155 @@
...
+
+In order to easily mock all actions done by a napalm driver, a driver `mock`
```
Comment (sieben) [inline docs/tutorials/mock_driver.rst]: A driver `mock` can mock all actions done by a common napalm driver.

File: docs/tutorials/mock_driver.rst [hunk 3]
```diff
@@ -1,30 +1,155 @@
...
+        hostname='foo', username='user', password='pass',
+        optional_args={
+            'path': path_to_results, 'profile': ('model_mocked', )
+        }
+    )
+
+Like other drivers, ``mock`` takes some optional arguments:
+
+- ``path`` - Required. Directory where results files are located
+- ``profile`` - Device type to mock: ``['ios']``, ``['nxos']``, ``['junos']``,
```
Comment (sieben) [inline docs/tutorials/mock_driver.rst]: Can you take it out if you don't use it? Otherwise, it's confusing.

**Comment by sieben** [comment]
`with` and context management helps to visualize better the lifecycle of a variable. Even if from a GC perspective it's not necessary an issue, it is still a good practice to see how long a variable will remain alive.
ok to remove the ansible part of the documentation. It can be added later.

**Comment by Anthony25** [comment]
Sure, I totally get why context management could be helpful. Just saying that it is not usable with napalm drivers yet.

**Comment by dbarrosop** [comment]
> Sure, I totally get why context management could be helpful. Just saying that it is not usable with napalm drivers yet.

It is. It's even documented:

http://napalm.readthedocs.io/en/latest/tutorials/context_manager.html

**Comment by dbarrosop** [comment]
Overall looks good. Would you mind adding a quick example mocking errors? You can see an example here: https://github.com/napalm-automation/napalm-ansible/blob/develop/tests/napalm_install_config/mocked/merge.error/load_merge_candidate.1

**Comment by Anthony25** [comment]
Oh sorry, I wasn't aware about context management in drivers. I'll simplify my doc by using it, thanks!

Thanks for the link about errors mocking. I was not sure if it was possible or not.

**Comment by dbarrosop** [comment]
No problem, thanks for helping out documenting this :)

**Comment by dbarrosop** [comment]
Is this ready? Are you happy with it?

**Comment by Anthony25** [comment]
Again, sorry for the delay. I added some documentations about error handling.

File: docs/tutorials/mock_driver.rst [hunk 4]
```diff
@@ -1,36 +1,177 @@
-Napalm mock driver
-==================
+Unit tests: Mock driver
+=======================
+
+A mock driver is a software that imitates the response pattern of another
+system. It is meant to do nothing but returns the same predictible result,
```
Comment (sieben) [inline docs/tutorials/mock_driver.rst]: predictable

File: docs/tutorials/mock_driver.rst [hunk 5]
```diff
@@ -1,36 +1,177 @@
...
+        "Ethernet1/1": {
+            "ipv6": {"2001:DB8::": {"prefix_length": 64}}
+        }
+    }
+
+Mock a CLI call
+~~~~~~~~~~~~~~~
+
+``device.cli(commands)`` calls are a bit different to mock, as a suffix
+corresponding to the command applied to the device need to be added. As before,
```
Comment (sieben) [inline docs/tutorials/mock_driver.rst]: needs

**Comment by dbarrosop** [comment]
Great work!

**Comment by dbarrosop** [comment]
could you open it against napalm repo? :D

**Pull request merged**